### PR TITLE
Cuda Profiler 

### PR DESCRIPTION
**Pull request by RandySheriffH**
Utilize cupti library to profile cuda kernels.
The stop and stop profiling are synced up with ongoing profiling mechanism.
Callbacks will collect kernel numbers asynchronously to minimize overhead.
After profiling, records will be added to existing ones with the same format:
{"cat" : "Kernel",
 "pid" :50556,
 "tid" :51460,
 "dur" :16,
 "ts" :104,
 "ph" : "X",
  "name":"cudnn_kernel_name....",
  "args" : {"stream":0, "block_x":..., "block_y":..., "block_z": ...}
}

### CODE REVIEW

File: cmake/onnxruntime_common.cmake [hunk 1]
```diff
@@ -79,6 +79,12 @@ source_group(TREE ${REPO_ROOT} FILES ${onnxruntime_common_src})
 
 add_library(onnxruntime_common ${onnxruntime_common_src})
 
+if (onnxruntime_USE_CUDA)
+  target_include_directories(onnxruntime_common PUBLIC ${onnxruntime_CUDA_HOME}/include ${onnxruntime_CUDA_HOME}/extras/CUPTI/include)
+  target_link_directories(onnxruntime_common PUBLIC ${onnxruntime_CUDA_HOME}/extras/CUPTI/lib64)
+  target_link_libraries(onnxruntime_common cupti)
+endif()
```
Comment (HectorSVC) [inline cmake/onnxruntime_common.cmake:86]: how much will this increase the package size?

File: tools/ci_build/github/windows/setup_env_cuda_11.bat [hunk 2]
```diff
@@ -1,2 +1,2 @@
-set PATH=C:\azcopy;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\bin;C:\local\cudnn-11.0-windows-x64-v8.0.2.39\cuda\bin;%PATH%
+set PATH=C:\azcopy;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\bin;C:\azcopy;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\extras\CUPTI\lib64;C:\local\cudnn-11.0-windows-x64-v8.0.2.39\cuda\bin;%PATH%
```
Comment (HectorSVC) [inline tools/ci_build/github/windows/setup_env_cuda_11.bat:1]: >C:\azcopy; [](start = 80, length = 10)

duplicate: C:\azcopy;

File: cmake/onnxruntime_common.cmake (hunk 1 above)
Comment (RandySheriffH) [reply cmake/onnxruntime_common.cmake:86]: Wondows cupti.lib 11.2 is 32 kb, similar for linux.

File: onnxruntime/core/common/profiler.cc [hunk 3]
```diff
@@ -2,11 +2,148 @@
...
+
+#ifdef USE_CUDA
+#include <cupti.h>
+#endif
 
 namespace onnxruntime {
 namespace profiling {
 using namespace std::chrono;
 
+class DeviceProfiler {
```
Comment (HectorSVC) [inline onnxruntime/core/common/profiler.cc:15]: >DeviceProfiler [](start = 6, length = 14)

should add some UT for this with CUDA build.

File: tools/ci_build/github/windows/setup_env_cuda_11.bat (hunk 2 above)
Comment (RandySheriffH) [reply tools/ci_build/github/windows/setup_env_cuda_11.bat]: Removed.

File: onnxruntime/core/common/profiler.cc (hunk 3 above)
Comment (RandySheriffH) [reply onnxruntime/core/common/profiler.cc:15]: Added.

**Review by HectorSVC — APPROVED** [review]
:shipit:

**Pull request merged**